### ISSUE TITLE
Recover value from Traced<ConcreteArray(-)>with<JVPTrace(level=1/0)>

### ISSUE THREAD

**Issue by LenCewa** [google/jax#2332]
Hi,
I'm using Jax to compute Fourier coefficients. I want to save the computed loss in every iteration. Therefore I do the following in my loss function:

```
    def loss(self, coefficents, train_values, train_labels):
        preds = self.batched_predict(coefficents, train_values)[:, 0]
        l = jnp.sum(jnp.square(preds - train_labels))
        self.loss_list += [l]
        return l
```

The loss functions get's called in my update function:

```
    def update(self, coefficients, x, y):
        grads = grad(self.loss)(coefficients, x, y)
        /*code*/
        return /*result*/
```
However, the list contains these Jax primitives:

`Traced<ConcreteArray(105.82802000294275)>with<JVPTrace(level=1/0)>`

I want to access the numbers (i.e. `105.82802000294275`) stored in my `loss_list` to plot them with pyplot and save the list of the loss values with numpy. Is there a way to "unwrap" the values in a Jax primitive?

**Comment by mattjj** [comment]
Thanks for the question!

Side-effects aren't allowed in JAX-transformed functions, and appending to a list like `self.loss_list += [l]` is a side-effect. You need all function outputs to be return values of the function.

Perhaps you can make use of [`value_and_grad`](https://jax.readthedocs.io/en/latest/jax.html#jax.value_and_grad) instead of `grad` here?

**Comment by LenCewa** [comment]
Alright, thanks for that quick response! I'll try it out.
